go/adbc/driver/snowflake/bulk_ingestion.go

```diff
@@ -0,0 +1,546 @@
+	// Maximum number of COPY operations to run concurrently.
+	//
+	// Bulk ingestion performance is optimized by executing COPY queries as files are
+	// still being uploaded. Snowflake COPY speed scales with warehouse size, so smaller
+	// warehouses may benefit from setting this value higher to ensure long-running
+	// COPY queries do not block newly uploaded files from being loaded.
+	// Default is 4. If set to 0, only a single COPY query will be executed as part of ingestion,
+	// once all files have finished uploading. Must be an integer >= 0.
+	copyConcurrency int
+	// Compression codec to use for Parquet files.
+	//
+	// When network speeds are high, it is generally faster to use a faster codec with
+	// a lower compression ratio. The opposite is true if the network is slow by CPU is
+	// available.
+	// Default is Snappy.
+	compressionCodec compress.Compression
+	// Compression level for Parquet files.
+	//
+	// The compression level is codec-specific. Some codecs do not support setting it,
+	// notably Snappy.
+	// Default is the default level for the specified compressionCodec.
+	compressionLevel int
+}
+
+func DefaultIngestOptions() *ingestOptions {
+	return &ingestOptions{
+		targetFileSize:    defaultTargetFileSize,
+		writerConcurrency: defaultWriterConcurrency,
+		uploadConcurrency: defaultUploadConcurrency,
+		copyConcurrency:   defaultCopyConcurrency,
+		compressionCodec:  defaultCompressionCodec,
+		compressionLevel:  defaultCompressionLevel,
+	}
+}
+
+// ingestRecord performs bulk ingestion of a single Record and returns the
+// number of rows affected.
+//
+// The Record must already be bound by calling stmt.Bind(), and will be released
+// and reset upon completion.
+func (st *statement) ingestRecord(ctx context.Context) (nrows int64, err error) {
+	defer func() {
+		// Record already released by writeParquet()
+		st.bound = nil
+	}()
+
+	parquetProps, arrowProps := newWriterProps(st.alloc, st.ingestOptions)
+	g := errgroup.Group{}
+
+	// writeParquet takes a channel of Records, but we only have one Record to write
+	recordCh := make(chan arrow.Record, 1)
+	recordCh <- st.bound
+	close(recordCh)
+
+	// Read the Record from the channel and write it into the provided buffer
+	schema := st.bound.Schema()
+	buf := new(bytes.Buffer)
+	g.Go(func() error {
+		err = writeParquet(schema, buf, recordCh, 0, parquetProps, arrowProps)
+		if err != io.EOF {
+			return err
+		}
+		return nil
+	})
+
+	// Create a temporary stage, we can't start uploading until it has been created
+	_, err = st.cnxn.cn.ExecContext(ctx, createTemporaryStageStmt, nil)
+	if err != nil {
+		return
+	}
+
+	// Wait for Parquet file to finish writing
+	err = g.Wait()
+	if err != nil {
+		return
+	}
+
+	// If successful, upload the file to Snowflake
+	fileName := "0.parquet" // Only writing 1 file, so use same name as first file written by ingestStream() for consistency
+	err = uploadStream(ctx, st.cnxn.cn, buf, fileName)
+	if err != nil {
+		return
+	}
```

Comment:
Good point. I updated this. I now have the writing and stage creation kick off at the same time so that the writer buffers can fill up for whatever time we're waiting on the stage query response. After that comes back the upload starts directly from the parquet file as it's being written.